feat: add List.toArrayMap

grunweg opened the pull request:
Calling List.toArray followed by Array.map currently implies a superfluous memory allocation: first, the list is converted to an array, then the array is mapped again. We can avoid this easily, by creating a new function List.toArrayMap.

Suggested by `kim-em` [on zulip](https://leanprover.zulipchat.com/#narrow/channel/144837-PR-reviews/topic/.2312190.20.60have.60.20vs.20.60let.60.20linter/near/486752051).

grunweg commented:
awaiting-review

grunweg commented:
@kim-em Here's the PR we discussed on zulip.

kim-em commented on Batteries/Data/List/ArrayMap.lean:
```diff
@@ -0,0 +1,49 @@
+-- def toArrayMapM {m : Type v → Type w} [Monad m] (l : List α) (f : α → m β) : m (Array β) :=
+--   l.foldlM (init := #[]) fun acc x => acc.push (f x)
+
+/-- info: #[3, 4, 5, 6] -/
+#guard_msgs in
+#eval toArrayMap [0, 1, 2, 3] (fun n => n + 3)
+
+/-- info: #[7, 9, 15, 25] -/
+#guard_msgs in
+#eval toArrayMap [0, 1, 2, 3] (fun n => 2 * n ^ 2 + 7)
```
These should go in the test/ folder.

mattrobball commented on Batteries/Data/List/ArrayMap.lean:
```diff
@@ -0,0 +1,49 @@
+-- def toArrayMapM {m : Type v → Type w} [Monad m] (l : List α) (f : α → m β) : m (Array β) :=
+--   l.foldlM (init := #[]) fun acc x => acc.push (f x)
+
+/-- info: #[3, 4, 5, 6] -/
```
Units tests should be in test files.

mattrobball reviewed (commented):
Can you close this and open a PR from a branch under the batteries repo? That should trigger the full CI suite.

grunweg commented:
I have rebased on the most recent master and split the unit tests into a separate file.

The README says to open PRs from a fork, though (hence I did so). I checked; I don't have permission to push to non-master branches of batteries. If this is insufficient, can you please either (a) tell me what else to do, (b) if there are missing permissions, change them, (c) update the instructions? Thanks a lot!

mattrobball commented:
Oh, perhaps I just needed to wait. Apologies. Once it builds mathlib, it should be good to go.

grunweg commented:
And mathlib builds :tada:
